lib/starttls.js

```diff
@@ -33,111 +15,46 @@ function starttlsClient(socket, options, callback) {
 }
 
 function starttls(socket, options, callback, isServer) {
-  var sslcontext;
-
   var opts = {};
+
   Object.keys(options).forEach(function(key) {
     opts[key] = options[key];
   });
   if (!opts.ciphers) {
     opts.ciphers = RECOMMENDED_CIPHERS;
   }
+  opts.isServer = isServer;
+  opts.secureContext = tls.createSecureContext(opts);
 
-  socket.removeAllListeners('data');
-  if (tls.createSecureContext) {
-    sslcontext = tls.createSecureContext(opts);
-  } else {
-    sslcontext = crypto.createCredentials(opts);
-  }
-  var pair = tls.createSecurePair(sslcontext, isServer);
-  var cleartext = pipe(pair, socket);
-
+  var secureSocket = new tls.TLSSocket(socket, opts);
   var erroredOut = false;
-  pair.on('secure', function() {
+
+  // NodeJS documentation bug: secure vs secureConnect
+  // https://github.com/nodejs/node/issues/10555
+  secureSocket.on('secure', function() {
     if (erroredOut) {
-      pair.end();
+      secureSocket.end();
       return;
     }
 
-    var verifyError = (pair._ssl || pair.ssl).verifyError();
-
-    if (verifyError) {
-      cleartext.authorized = false;
-      cleartext.authorizationError = verifyError;
+    var authError = secureSocket.ssl.verifyError();
+    if (authError) {
+        secureSocket.authorized = false;
+        secureSocket.authorizationError = authError;
     } else {
-      cleartext.authorized = true;
+        secureSocket.authorized = true;
```

Comment:
I'm not sure I can do that as I don't know the hostname of client. This is server side of FTP connection, so client is connecting to us. All we can know about the client is perhaps its IP address and even that one does not have to be correct (if client is behind NAT or so). So making sure that the certificate presented by the client is valid is probably the best we can do here. It looks that neither the previous version of the code called checkServerIdentity(). Since it's not a regression I would prefer to leave the code as it is.